core/lib/dal/src/events_dal.rs

```diff
@@ -217,9 +234,169 @@ impl EventsDal<'_, '_> {
+                                event_index_in_block ASC
+                        ) - 1 AS event_index_in_block_without_eth_transfer,
+                        ROW_NUMBER() OVER (
+                            PARTITION BY
+                                miniblock_number,
+                                tx_hash
+                            ORDER BY
+                                event_index_in_tx ASC
+                        ) - 1 AS event_index_in_tx_without_eth_transfer
+                    FROM
+                        events
+                    WHERE
+                        miniblock_number BETWEEN $1 AND $2
+                        AND NOT (
+                            address = $3
+                            AND topic1 = $4
+                        )
+                        AND (
+                            event_index_in_tx_without_eth_transfer IS NULL
+                            OR event_index_in_block_without_eth_transfer IS NULL
+                        )
+                ) AS subquery
+            WHERE
+                events.miniblock_number = subquery.miniblock_number
+                AND events.tx_hash = subquery.tx_hash
+                AND events.tx_index_in_block = subquery.tx_index_in_block
+                AND events.event_index_in_block = subquery.event_index_in_block
+                AND events.event_index_in_tx = subquery.event_index_in_tx
+            "#,
+            numbers.start().0 as i64,
+            numbers.end().0 as i64,
+            L2_ETH_TOKEN_ADDRESS.as_bytes(),
+            TRANSFER_EVENT_TOPIC.as_bytes()
+        )
+        .execute(self.storage.conn())
+        .await?;
+
+        Ok(result.rows_affected() + count_eth_transfer_events)
+    }
+
+    /// Method that assigns 0 for all the events with ETH transfer.
+    async fn assign_eth_transfer_event_indexes(
+        &mut self,
+        numbers: ops::RangeInclusive<MiniblockNumber>,
+    ) -> sqlx::Result<u64> {
+        let result = sqlx::query!(
+            r#"
+            UPDATE events
+            SET
+                event_index_in_block_without_eth_transfer = 0,
+                event_index_in_tx_without_eth_transfer = 0
+            WHERE
+                miniblock_number BETWEEN $1 AND $2
+                AND address = $3
+                AND topic1 = $4
+                AND (
+                    event_index_in_tx_without_eth_transfer IS NULL
+                    OR event_index_in_block_without_eth_transfer IS NULL
+                )
+            "#,
+            numbers.start().0 as i64,
+            numbers.end().0 as i64,
+            L2_ETH_TOKEN_ADDRESS.as_bytes(),
+            TRANSFER_EVENT_TOPIC.as_bytes()
+        )
+        .execute(self.storage.conn())
+        .await?;
+
+        Ok(result.rows_affected())
+    }
+
+    /// Method checks, whether any event in the miniblock has `index_without_eth_transfer` equal to NULL, which means that indexes are not migrated.
+    pub async fn are_event_indexes_migrated(
```

Comment:
Is it strictly necessary to check all miniblocks in the range? Note that in #688, it's sufficient to check only the starting miniblock of the range provided that ranges are guaranteed to be identical across node restarts. Would a similar assumption not apply here? AFAIU, checking the entire range would put significantly more load on the DB if the range is reasonably large.